tower-http/src/set_response_header.rs

```diff
@@ -0,0 +1,271 @@
+//!     Ok::<_, Infallible>(Response::new(request.into_body()))
+//! });
+//!
+//! let mut svc = ServiceBuilder::new()
+//!     .layer(
+//!         // Layer that sets `Content-Type: text/html` on responses.
+//!         //
+//!         // We have to add `::<_, Body>` since Rust cannot infer the body type when
+//!         // we don't use a closure to produce the header value.
+//!         SetResponseHeaderLayer::<_, Body>::new(
+//!             HeaderName::from_static("content-type"),
+//!             HeaderValue::from_static("text/html"),
+//!         )
+//!     )
+//!     .layer(
+//!         // Layer that sets `Content-Length` if the body has a known size.
+//!         // Bodies with streaming responses wont have a known size.
+//!         SetResponseHeaderLayer::new(
+//!             HeaderName::from_static("content-length"),
+//!             |response: &Response<Body>| {
+//!                 if let Some(size) = response.body().size_hint().exact() {
+//!                     Some(HeaderValue::from_str(&size.to_string()).unwrap())
+//!                 } else {
+//!                     None
+//!                 }
+//!             }
+//!         )
+//!     )
+//!     .service(echo_service);
+//!
+//! let request = Request::new(Body::from("<strong>Hello, World</strong>"));
+//!
+//! let response = svc.ready_and().await?.call(request).await?;
+//!
+//! assert_eq!(response.headers()["content-type"], "text/html");
+//! assert_eq!(response.headers()["content-length"], "29");
+//! #
+//! # Ok(())
+//! # }
+
+use futures_util::ready;
+use http::{header::HeaderName, HeaderValue, Request, Response};
+use pin_project::pin_project;
+use std::{
+    fmt,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use std::{future::Future, marker::PhantomData};
+use tower_layer::Layer;
+use tower_service::Service;
+
+/// Layer that applies [`SetResponseHeader`] which adds a response header.
+///
+/// See [`SetResponseHeader`] for more details.
+pub struct SetResponseHeaderLayer<M, B> {
+    header_name: HeaderName,
+    make: M,
+    override_existing: bool,
+    _marker: PhantomData<fn() -> B>,
+}
+
+impl<M, B> fmt::Debug for SetResponseHeaderLayer<M, B> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("SetResponseHeaderLayer")
+            .field("header_name", &self.header_name)
+            .field("override_existing", &self.override_existing)
+            .field("make", &std::any::type_name::<M>())
+            .finish()
+    }
+}
+
+impl<M, B> SetResponseHeaderLayer<M, B> {
+    /// Create a new [`SetResponseHeaderLayer`].
+    pub fn new(header_name: HeaderName, make: M) -> Self
+    where
+        M: MakeHeaderValue<B>,
+    {
+        Self {
+            make,
+            header_name,
+            override_existing: true,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Should the header be overriden if the response already contains it?
```

Comment:
also, since HTTP headers can have multiple values, I wonder if it's worth having a third "mode" for this, which calls [`HeaderMap::append`](https://docs.rs/http/0.2.3/http/header/struct.HeaderMap.html#method.append) instead of `insert`, which adds a new value for a header regardless of whether any previous values are present or not? that way, there's also a mechanism to add _additional_ values to a header.

(FWIW, an append mode is probably particularly useful for proxies --- you might be forwarding a response from an upstream server, and you want to add your own values to the headers set by that upstream. For example, a response might have come via a _different_ proxy, and already have a `Via: 1.1 router-proxy` or something, and I might want to add `Via: 1.1 tls-proxy`, or whatever)

Reply:
Yeah thats a good idea. I kinda think it makes sense to change the API a bit then. I would find it a bit odd to have `override_existing` and `append`. Those two are mutually exclusive in a way. I changed this a bit to instead take an `InsertHeaderMode` enum.

Let me know what you think 😊